include/boost/numeric/odeint/stepper/controlled_adams_bashforth_moulton.hpp

```diff
@@ -39,39 +39,54 @@ class default_order_adjuster
     : m_algebra( algebra )
     {};
 
-    void adjust_order(size_t &order, const boost::array<wrapped_state_type, 3> & xerr)
+    void adjust_order(size_t &order, size_t init, boost::array<wrapped_state_type, 4> &xerr)
     {
         using std::abs;
 
-        value_type errm = abs(m_algebra.norm_inf(xerr[0].m_v));
-        value_type errc = abs(m_algebra.norm_inf(xerr[1].m_v));
-        value_type errp = abs(m_algebra.norm_inf(xerr[2].m_v));
+        value_type errc = abs(m_algebra.norm_inf(xerr[2].m_v));
 
-        if(order == 1)
-        {   
-            if(errp < errc)
-            {
-                order ++;
-            }
+        value_type errm1 = 3*errc;
+        value_type errm2 = 3*errc;
+
+        if(order > 2)
+        {
+            errm2 = abs(m_algebra.norm_inf(xerr[0].m_v));
         }
-        else if(order == MaxOrder)
+        if(order >= 2)
         {
-            if(errm < errc)
-            {
-                order --;
-            }
+            errm1 = abs(m_algebra.norm_inf(xerr[1].m_v));
         }
-        else
+
+        size_t o_new = order;
+
+        if(order == 2 && errm1 <= 0.5*errc)
+        {
+            o_new = order - 1;
+        }
+        else if(order > 2 && errm2 < errc && errm1 < errc)
+        {
+            o_new = order - 1;
+        }
+
+        if(init < order)
+        {
+            ++order;
+        }
+        else if(o_new == order -1)
+        {
+            --order;
+        }
+        else if(order < MaxOrder)
         {
-            if(errm < errc && 
-                errm < errp)
+            value_type errp = abs(m_algebra.norm_inf(xerr[3].m_v));
+
+            if(order > 1 && errm1 < errc && errp)
             {
-                order--;
+                --order;
```

Comment:
this branch is only executed if order < MaxOrder, is this right? Feels like this should also be applied if order == MaxOrder?

Reply:
This is absolutely right.